ipa-core/src/protocol/ipa_prf/malicious_security/lagrange.rs

```diff
@@ -0,0 +1,329 @@
+        let mut e_y = vec![Fp25519::ONE; e_p_size];
+        lagrange_evaluation(&e_p, &q, &e_p, &mut e_y);
+        assert_eq!(e_y, q);
+    }
+
+    #[test]
+    fn consistency_with_monomial_base_polynomial_test() {
+        // use random evaluation points
+        let mut rng = thread_rng();
+        let e_p_size = rng.gen::<usize>() % 20;
+        let mut e_p = vec![Fp25519::ONE; e_p_size];
+        e_p.iter_mut().for_each(|x| *x = rng.gen::<Fp25519>());
+
+        // sample random polynomial p in monomial form, i.e. `sum coefficient_k * x^k`,
+        // of degree `evaluation_points_size`
+        let mut p = vec![Fp25519::ONE; e_p_size];
+        p.iter_mut().for_each(|x| *x = rng.gen::<Fp25519>());
+        // add random point to evaluation points
+        e_p.push(rng.gen::<Fp25519>());
+
+        // evaluate polynomial p at evaluation_points and random point using monomial base
+        let mut y_values = vec![Fp25519::ZERO; e_p_size + 1];
+        for x_value in &e_p {
+            // monomial base, i.e. `x^k`
+            let mut base = Fp25519::ONE;
+            // evaluate p via `sum coefficient_k * x^k`
+            for coefficient in &p {
+                y_values.iter_mut().for_each(|y| *y = *coefficient * base);
+                base *= *x_value;
+            }
+        }
+
+        // lagrange evaluate p at random point
+        let mut e_y = Fp25519::ONE;
+        lagrange_evaluation(
+            &e_p[0..e_p_size],
+            &y_values[0..e_p_size],
+            std::slice::from_ref(&e_p[e_p_size]),
+            std::slice::from_mut(&mut e_y),
+        );
+
+        // check equality between "monomial evaluation" and "Lagrange evaluation"
+        assert_eq!(y_values[e_p_size], e_y);
+    }
+
+    #[test]
+    fn lagrange_precomputed_canonical_test() {
+        let mut rng = thread_rng();
+
+        // generate canonical evaluation points
+        let e_p_size = rng.gen::<usize>() % 20;
+        let mut e_p = vec![Fp25519::ONE; e_p_size];
+        generate_evaluation_points(&mut e_p);
+
+        // random polynomial p
+        let mut q = vec![Fp25519::ONE; e_p_size];
+        q.iter_mut().for_each(|y| *y = rng.gen::<Fp25519>());
+
+        // generate random points
+        let r_p_size = rng.gen::<usize>() % 20;
+        let mut r_p = vec![Fp25519::ONE; r_p_size];
+        r_p.iter_mut().for_each(|y| *y = rng.gen::<Fp25519>());
+
+        // non-precomputed
+        let mut evaluated_q = vec![Fp25519::ONE; r_p_size];
+        lagrange_evaluation(&e_p, &q, &r_p, &mut evaluated_q);
+
+        // precomputed
+        let mut evaluated_q_precomputed = vec![Fp25519::ONE; r_p_size];
+        let mut denominators = vec![Fp25519::ONE; e_p_size];
+        compute_lagrange_denominator(&e_p, &mut denominators);
+        let mut base = vec![vec![Fp25519::ONE; e_p_size]; r_p_size];
+        compute_lagrange_base(&r_p, &e_p, &denominators, &mut base);
+        lagrange_evaluation_precomputed(&q, &base, &mut evaluated_q_precomputed);
```

Comment:
An alternative approach, instead of just comparing the output of two implementations of lagrange_evaluation, would be to generate a random polynomial (i.e. generate random coefficients), and sample the evaluation points and the test points by just evaluating the polynomial, then to compare the test points with the values obtained via lagrange interpolation.